stan/math/rev/fun/append_col.hpp

```diff
@@ -0,0 +1,153 @@
+#ifndef STAN_MATH_REV_FUN_APPEND_COL_HPP
+#define STAN_MATH_REV_FUN_APPEND_COL_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/append_col.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the result of appending the second argument matrix after the
+ * first argument matrix, that is, putting them side by side, with
+ * the first matrix followed by the second matrix.
+ *
+ * Given input types result in following outputs:
+ * (matrix, matrix) -> matrix,
+ * (matrix, vector) -> matrix,
+ * (vector, matrix) -> matrix,
+ * (vector, vector) -> matrix,
+ * (row vector, row vector) -> row_vector.
+ *
+ * @tparam T1 A `var_value` with inner matrix type.
+ * @tparam T1 A `var_value` with inner matrix type.
+ *
+ * @param A First matrix.
+ * @param B Second matrix.
+ * @return Result of appending the first matrix followed by the
+ * second matrix side by side.
+ */
+template <typename T1, typename T2, require_any_var_matrix_t<T1, T2>* = nullptr>
+inline auto append_col(const T1& A, const T2& B) {
+  check_size_match("append_col", "columns of A", A.rows(), "columns of B",
+                   B.rows());
+  if (!is_constant<T1>::value && !is_constant<T2>::value) {
+    arena_t<promote_scalar_t<var, T1>> arena_A = A;
+    arena_t<promote_scalar_t<var, T2>> arena_B = B;
+    return make_callback_var(
+        append_col(value_of(arena_A), value_of(arena_B)),
+        [arena_A, arena_B](auto& vi) mutable {
+          arena_A.adj() += vi.adj().leftCols(arena_A.cols());
+          arena_B.adj() += vi.adj().rightCols(arena_B.cols());
+        });
+  } else if (!is_constant<T1>::value) {
+    arena_t<promote_scalar_t<var, T1>> arena_A = A;
+    arena_t<promote_scalar_t<double, T2>> arena_B = value_of(B);
```

Comment:
We don't need this. `B` could be directly used on the next line.

There are more examples of the same thing in this PR.